src/sys/windows.rs

```diff
@@ -92,29 +90,31 @@ pub fn reflink(from: &Path, to: &Path) -> io::Result<()> {
             debug_assert_eq!(bytes_copied % cluster_size, 0);
         }
         unsafe {
-            *dup_extent.SourceFileOffset.QuadPart_mut() = bytes_copied;
-            *dup_extent.TargetFileOffset.QuadPart_mut() = bytes_copied;
-            *dup_extent.ByteCount.QuadPart_mut() = bytes_to_copy;
+            *(*dup_extent.as_mut_ptr()).SourceFileOffset.QuadPart_mut() = bytes_copied;
+            *(*dup_extent.as_mut_ptr()).TargetFileOffset.QuadPart_mut() = bytes_copied;
+            *(*dup_extent.as_mut_ptr()).ByteCount.QuadPart_mut() = bytes_to_copy;
         }
         let mut bytes_returned = 0u32;
         let res = unsafe {
             DeviceIoControl(
                 dest.as_raw_handle() as _,
                 ffi::FSCTL_DUPLICATE_EXTENTS_TO_FILE,
-                &mut dup_extent as *mut _ as *mut _,
-                mem::size_of::<ffi::DUPLICATE_EXTENTS_DATA>() as u32,
+                dup_extent.as_mut_ptr() as *mut _,
+                mem::size_of::<ffi::DUPLICATE_EXTENTS_DATA>()
+                    .try_into()
+                    .unwrap(),
                 ptr::null_mut(),
                 0,
                 &mut bytes_returned as *mut _,
                 ptr::null_mut(),
             )
         };
         if res == 0 {
-            let _ = fs::remove_file(to);
             return Err(io::Error::last_os_error());
         }
         bytes_copied += bytes_to_copy;
```

Comment:
I'm not sure this is correct.
Perhaps it should be

```suggestion
        bytes_copied += bytes_returned;
```

@passcod Can you verify that this piece of code is correct as I have zero experience with windows?

Reply:
Right, so this is correct as is.

[This API](https://docs.microsoft.com/en-us/windows/win32/api/winioctl/ni-winioctl-fsctl_duplicate_extents_to_file)'s last argument is `lpOverlapped`, and here it's passed as null. The [documentation](https://docs.microsoft.com/en-us/windows-hardware/drivers/ifs/fsctl-duplicate-extents-to-file-ex) also states that `lpOutBuffer` is not used for block cloning (which is [the friendly name](https://docs.microsoft.com/en-us/windows/win32/fileio/block-cloning) for this `FSCTL_DUPLICATE_EXTENTS_TO_FILE` thing). And so:

> If `lpOverlapped` is NULL, `lpBytesReturned` cannot be NULL. Even when an operation returns no output data and `lpOutBuffer` is NULL, [`DeviceIoControl`](https://docs.microsoft.com/en-us/windows/win32/api/ioapiset/nf-ioapiset-deviceiocontrol) makes use of `lpBytesReturned`. **After such an operation, the value of `lpBytesReturned` is meaningless.**

(emphasis mine)

so we must provide its `&mut` to the API, but its value after the API call is done should not be used.

Reply:
What about the return value?

Is `bytes_to_copy` equal to `bytes_returned`?

Reply:
THanks for the verification!